src/io/file_reader.hpp

```diff
@@ -0,0 +1,49 @@
+#pragma once
+#include <cassert>
+#include <cstdio>
+#include <vector>
+#include <exceptions/exceptions.hpp>
+
+namespace docgen {
+namespace io {
+
+struct file_reader
+{
+    file_reader(const char* filename)
+        : file_(fopen(filename, "r"))
```

Comment:
I'd like to change this to just take a `FILE *` and hold it, rather than deal with open/close. This system-level interfacing probably shouldn't be handled within the lexer anyway—it should just take a stream and run with it.